presto-main/src/test/java/com/facebook/presto/operator/TestPartitionedOutputOperator.java

```diff
@@ -0,0 +1,274 @@
+            partitionedOutputOperator.addInput(createSimplePage());
+        }
+        partitionedOutputOperator.finish();
+        OperatorContext operatorContext = partitionedOutputOperator.getOperatorContext();
+        long outputSize = operatorContext.getOutputDataSize().getTotalCount();
+        long outputPositionCount = operatorContext.getOutputPositions().getTotalCount();
+        assertEquals(outputSize, 900_000);
+        assertEquals(outputPositionCount, 100_000);
+        assertEquals(createSimplePage().getSizeInBytes() * NUM_OF_ITERATIONS, outputSize);
+    }
+
+    @Test
+    public void testAddInputForPageWithDictionary()
+    {
+        PartitionedOutputOperator partitionedOutputOperator = createPartitionedOutputOperator(false);
+        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
+            partitionedOutputOperator.addInput(createPageWithDictionary());
+        }
+        partitionedOutputOperator.finish();
+        OperatorContext operatorContext = partitionedOutputOperator.getOperatorContext();
+        long outputSize = operatorContext.getOutputDataSize().getTotalCount();
+        long outputPositionCount = operatorContext.getOutputPositions().getTotalCount();
+        assertEquals(outputSize, 900_000);
+        assertEquals(outputPositionCount, 100_000);
+        assertTrue(outputSize > createPageWithDictionary().getSizeInBytes() * NUM_OF_ITERATIONS);
+    }
+
+    @Test
+    public void testAddInputForPageWithRunLength()
+    {
+        PartitionedOutputOperator partitionedOutputOperator = createPartitionedOutputOperator(false);
+        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
+            partitionedOutputOperator.addInput(createPageWithRunLength());
+        }
+        partitionedOutputOperator.finish();
+        OperatorContext operatorContext = partitionedOutputOperator.getOperatorContext();
+        long outputSize = operatorContext.getOutputDataSize().getTotalCount();
+        long outputPositionCount = operatorContext.getOutputPositions().getTotalCount();
+        assertEquals(outputSize, 900_000);
+        assertEquals(outputPositionCount, 100_000);
+        assertTrue(outputSize > createPageWithRunLength().getSizeInBytes() * NUM_OF_ITERATIONS);
+    }
+
+    @Test
+    public void testAddInputForSimplePageAndReplication()
+    {
+        PartitionedOutputOperator partitionedOutputOperator = createPartitionedOutputOperator(true);
+        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
+            partitionedOutputOperator.addInput(createSimplePage());
+        }
+        partitionedOutputOperator.finish();
+        OperatorContext operatorContext = partitionedOutputOperator.getOperatorContext();
+        long outputSize = operatorContext.getOutputDataSize().getTotalCount();
+        long outputPositionCount = operatorContext.getOutputPositions().getTotalCount();
+        assertEquals(outputSize, 904599);
+        assertEquals(outputPositionCount, 100_511);
```

Comment:
why is it 100_511? Could you please verify if all rows were indeed replicated and copied to all pageBuilders?

Reply:
Math behind this .
100_511 = 1511 + 1000*(100-1)
Order of progression is this.
|  Num of Iterations | Position Count  |
|--| -|
| 1  | 1511  |
|  2 | 2511  |
|   3|  3511 |
|  4 | 4511  |
|  10 | 10511  |
|  100 | 100511  |

Was able to debug through the code and check that all appendRow was getting called for all PageBuilders. Let me know if this looks good

Reply:
You need to add a nullChannel in the page in order for the replicated case to happen. Let's see this piece of code in partitionPage() to understand how we can reproduce the case every row is replicated:

```
boolean shouldReplicate = (replicatesAnyRow && !hasAnyRowBeenReplicated) ||
                        nullChannel.isPresent() && page.getBlock(nullChannel.getAsInt()).isNull(position);
```
You set the replicatesAnyRow to true when constructing the operator, but it only replicates one row and the next time hasAnyRowBeenReplicated would be true and no more rows would be replicated. If we want to replicate all rows, it needs nullChannel.isPresent() && page.getBlock(nullChannel.getAsInt()).isNull(position) to be true. For that we need an extra block that contains all null values. We can construct such a page like this:

```
private Page createPageWithNullBlock()
    {
        Block[] blocks = new Block[2];
        blocks[0] = new RunLengthEncodedBlock(BIGINT.createBlockBuilder(null, 1).appendNull().build(), POSITION_COUNT);
        blocks[1] = buildBigintBlock(POSITION_COUNT, false);
        
        return new Page(POSITION_COUNT, blocks);
    }
```
Then in the replicated test cases, you add this page instead of the simple page. The expected output number of rows should be 51,200,000.

Reply:
A quick note here. To properly account for the sizeInBytes the PartitionOperator construction has to change to account for the null block addition for replicated scenarios.
I have added that as well. Please review.